File: backend/payment/src/main/java/com/bbmovie/payment/service/zalopay/ZaloHmacUtil.java
package com.bbmovie.payment.service.zalopay;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

@SuppressWarnings({"squid:S1118", "unused"})
public final class ZaloHmacUtil {
    public static final String HMACMD5 = "HmacMD5";
    public static final String HMACSHA1 = "HmacSHA1";
    public static final String HMACSHA256 = "HmacSHA256";
    public static final String HMACSHA512 = "HmacSHA512";

    private static byte[] hmacEncode(String algorithm, String key, String data) {
        try {
            Mac mac = Mac.getInstance(algorithm);
            mac.init(new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), algorithm));
            return mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
        } catch (Exception e) {
            return new byte[0];
        }
    }

    public static String hmacHexStringEncode(String algorithm, String key, String data) {
        byte[] result = hmacEncode(algorithm, key, data);
        return result != null ? ZaloHexStringUtil.byteArrayToHexString(result) : null;
    }
}

File: backend/payment/src/main/kotlin/com/bbmovie/payment/service/payment/zalopay/ZaloHmacUtil.kt
package com.bbmovie.payment.service.payment.zalopay

import java.util.LinkedList
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

@Suppress("unused")
object ZaloHmacUtil {

    const val HMACMD5 = "HmacMD5"
    const val HMACSHA1 = "HmacSHA1"
    const val HMACSHA256 = "HmacSHA256"
    const val HMACSHA512 = "HmacSHA512"
    val UTF8CHARSET = Charsets.UTF_8
    val HMACS = LinkedList(listOf("UnSupport", "HmacSHA256", "HmacMD5", "HmacSHA384", "HMacSHA1", "HmacSHA512"))

    private fun hmacEncode(algorithm: String, key: String, data: String): ByteArray? {
        val macGenerator = try {
            Mac.getInstance(algorithm).apply {
                init(SecretKeySpec(key.toByteArray(UTF8CHARSET), algorithm))
            }
        } catch (ex: Exception) {
            null
        }

        val dataByte = data.toByteArray(UTF8CHARSET)
        return macGenerator?.doFinal(dataByte)
    }

    fun hmacHexStringEncode(algorithm: String, key: String, data: String): String? {
        return hmacEncode(algorithm, key, data)?.let {
            ZaloHexStringUtil.byteArrayToHexString(it)
        }
    }
}